Go config hooks

tonyespy opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [NA] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## What is the current behavior?
This PR adds go-based configure and install hooks, including support for configuration overrides. Please see the documentation in `snap/README.md file`.

siggiskulason commented on hooks/cmd/configure/configure.go:
```diff
@@ -0,0 +1,261 @@
+		serviceCfg := hooks.EnvConfig + "." + v
+		envJSON, err = cli.Config(serviceCfg)
+		if err != nil {
+			return errors.New(fmt.Sprintf("edgexfoundry:configure failed to read service %s configuration - %v", v, err))
```
could replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...)

siggiskulason commented on hooks/const.go:
```diff
@@ -0,0 +1,163 @@
+package hooks
+
+const (
+	AutostartConfig = "autostart"
```
As these constants are exported, they should have a comment

siggiskulason commented on hooks/const.go:
```diff
@@ -0,0 +1,163 @@
+	"add-secretstore-tokens": "ADD_SECRETSTORE_TOKENS",
+}
+
+var Services = []string{
```
"Services" is exported and should have a comment

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+var (
+	debug        bool = false
+	log          *syslog.Writer
+	Snap         string
```
The Snap.. strings are exported and should have a comment

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+	return nil
+}
+
+// CopyFile copies a file within the snap
```
should be "CopyFileReplace copies..."

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+	return nil
+}
+
+func Debug(msg string) {
```
should have comment or be unexported

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+	}
+}
+
+func Error(msg string) {
```
should have comment or be unexported

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+	log.Err(msg)
+}
+
+func Info(msg string) {
```
should have comment or be unexported

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+	return nil
+}
+
+func InitLog(setDebug bool) error {
```
InitLog should have comment or be unexported

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+//	return &Client{getter: g}
+//}
+
+// Get uses snapctl to get a value from a key, or returns error
```
Update comment to Config..

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+	return strings.TrimSpace(string(out)), nil
+}
+
+// Get uses snapctl to set a config value from a key, or returns error
```
Update comment to SetConfig..

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+
+	err := exec.Command("snapctl", "set", fmt.Sprintf("%s=%s", key, val)).Run()
+	if err != nil {
+		return errors.New(fmt.Sprintf("snapctl SET failed for %s - %v", key, err))
```
replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...)

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+
+	err := cmd.Run()
+	if err != nil {
+		return errors.New(fmt.Sprintf("snapctl start %s failed - %v", name, err))
```
replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...)

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+
+	err := cmd.Run()
+	if err != nil {
+		return errors.New(fmt.Sprintf("snapctl stop %s failed - %v", name, err))
```
replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...)

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+
+	err := json.Unmarshal([]byte(envJSON), &m)
+	if err != nil {
+		return errors.New(fmt.Sprintf("failed to unmarshall EdgeX config - %v", err))
```
replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...)

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+	for k, v := range flatConf {
+		env, ok := ConfToEnv[k]
+		if !ok {
+			Error(fmt.Sprintf("invalid EdgeX config option - %s", k))
```
As this is a non-fatal error, should it be an error or a warning?

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+
+	err = ioutil.WriteFile(tmp, b.Bytes(), 0644)
+	if err != nil {
+		return errors.New(fmt.Sprintf("failed to write service.env file - %v", err))
```
replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...)

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+
+	err = os.Rename(tmp, path)
+	if err != nil {
+		return errors.New(fmt.Sprintf("failed to rename service.env.tmp file - %v", err))
```
replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...)

siggiskulason commented on hooks/utils.go:
```diff
@@ -0,0 +1,327 @@
+
+	// TODO: is this approach sufficient, or should this function be re-written to
+	// use file locking, or minimally run f.Sync()?
+
```
Did you come to a conclusion on if syncing is needed?

siggiskulason commented on snap/README.md:
````diff
@@ -125,6 +125,30 @@ Or by using the systemd unit name and `journalctl`:
 ```
 
+### Configuration Overrides
+The EdgeX snap support configuration overrides via its configure and install hooks which generate service-specific .env files
````
support -> supports

siggiskulason reviewed (commented):
Excellent implementation. I just had some minor comments and will now merge my kong prototype code into this.

tonyespy commented on cmd/security-secrets-setup/res/pkisetup-kong.json:
```diff
@@ -18,7 +18,7 @@
         "ca_o": "EdgeXFoundry"
     },
     "x509_tls_server_parameters": {
-        "tls_host": "edgex-kong",
```
I'll rebase the corresponding commit with a better explanation of why the change was made.  Yes, it's supposed to be set to 'localhost' instead of 'edgex-kong'.

Since we don't know what machine the snap we chose to use localhost for the TLS certificate for Kong. The shell-based install hook directly edited this file before moving it into place in $SNAP_DATA/config/security-secrets-setup/res. Instead of replicating that code, patching the file directly was a simplification.

tonyespy replied:
I just force pushed a rebase with a revised commit message.

tonyespy commented on hooks/cmd/configure/configure.go:
```diff
@@ -0,0 +1,261 @@
+	"fmt"
+	"os"
+
+	"github.com/canonical/app-service-configurable/hooks"
```
nice catch, this needs to be updated to point to edgex-go.

tonyespy commented:
Thanks for the preliminary review! I just addressed all of your comments and force-pushed.

siggiskulason commented:
Thanks for the changes! One other thing - you will need to rebase/merge in the changes from the commits for a) the update to versioning and b) the changes to enable the new secrets-config command.

tonyespy commented:
Just rebased and force-pushed to pickup the versioning change and the new secrets-config command.

tonyespy commented:
After more testing today I just force pushed another set of changes based on the new checkbox test I added to test the configure hook:

https://code.launchpad.net/~awe/+git/checkbox-provider-edgex/+merge/401656

This test validates the .env files and compares them to a set of test files to ensure the keys are all formatted correctly. The test does *not* currently restart any of the services and validate config overrides.  I've performed some manual testing that restarts work properly including:

  - core-data
  - device-virtual
  - security-secretstore-setup

Note, I tried to test security-proxy-setup, however it's currently not possible to restart it due to the fact that a Kong TLS has already been configured, thus setting env.security-proxy.add-proxy-route will only work from the install hook.

I also have *not* tested the install hook default snap settings in an image yet. I plan to do this tomorrow. Since the install hook shares code with the configure hook, fingers cross that this doesn't uncover any additional problems.

Given all of the above, I'm taking this PR out of draft status.

anonymouse64 commented on hooks/cmd/configure/configure.go:
```diff
@@ -0,0 +1,260 @@
+			return fmt.Errorf("edgexfoundry:configure failed to read service %s configuration - %v", v, err)
+		}
+
+		if envJSON != "" {
```
some comments about this would be helpful, AIUI what you are doing for every service is:
1) get the service's config, apply it to the env files
2) start or stop the service and it's dependencies if it needs to be started/stopped

is that correct ?

tonyespy replied:
This code only handles automatic creation of service-specific .env files, there's no automatic restarts happening, and in fact the snap/README explicitly calls this out. Given the existing behaviors (both EdgeX and snapd) around restarting services, I decided to be conservative and force the user to restart services if configuration is applied after install...

anonymouse64 commented on hooks/cmd/configure/configure.go:
```diff
@@ -0,0 +1,260 @@
+)
+
+const (
+	on  = "on"
```
these should be capitalized to make clear in the code they are global constants

tonyespy replied:
hahaha.... my initial impl did have them uppercase, but then I lower-cased them as they're only supposed to be global within the current package. If uppercase makes more sense, I'll change 'em back.

anonymouse64 commented on hooks/cmd/configure/configure.go:
```diff
@@ -0,0 +1,260 @@
+
+			// if we are turning kuiper on, make sure
+			// app-service-configurable is on too
+			if err = handleService("app-service-configurable", status); err != nil {
```
do we need the inverse order for stopping and starting? i.e. if app-service-configurable depends on kuiper, should we start kuiper first then start kuiper and do the inverse for stopping? or does app-service-configurable know how to wait for it's dependency properly?

tonyespy replied:
This is an exact port of the existing bash logic and I have yet to run into any issues to-date. I think with respect to Kuiper and ASC, we're OK, as ASC merely acts as a proxy on the message bus for Kuiper, so there's no tight coupling between the two...

anonymouse64 commented on hooks/cmd/configure/configure.go:
```diff
@@ -0,0 +1,260 @@
+			// the security-proxy consists of the following base services
+			// - kong
+			// - postgres (because kong requires it)
+			if err = handleService("postgres", status); err != nil {
```
same question as above, do we need different start/stop ordering here?

anonymouse64 commented on hooks/cmd/configure/configure.go:
```diff
@@ -0,0 +1,260 @@
+				}
+
+				// turn off security-secrets-setup too
+				if err = handleService("security-secrets-setup", off); err != nil {
```
I would imagine we would stop this first, then stop the security-proxy-setup, then stop kong, then stop postgres

anonymouse64 commented on hooks/cmd/configure/configure.go:
```diff
@@ -0,0 +1,260 @@
+		debug = true
+	}
+
+	err = hooks.InitLog(debug)
+	if err != nil {
```
nitpick, this can be combined into one statement inside the if condition

tonyespy replied:
Yea, I've used that construct in a bunch of places, and probably should do so by default everywhere...

anonymouse64 commented on hooks/cmd/configure/configure.go:
```diff
@@ -0,0 +1,260 @@
+	}
+
+	err = hooks.GetEnvVars()
+	if err != nil {
```
same

anonymouse64 commented on hooks/cmd/configure/configure.go:
```diff
@@ -0,0 +1,260 @@
+	hooks.Debug("edgexfoundry:configure hook running")
+
+	err = handleServices()
+	if err != nil {
```
same

anonymouse64 commented on hooks/cmd/install/install.go:
```diff
@@ -0,0 +1,360 @@
+		}
+
+		err := os.MkdirAll(destDir, 0755)
+		if err != nil {
```
same

anonymouse64 commented on hooks/cmd/install/install.go:
```diff
@@ -0,0 +1,360 @@
+		destPath := destDir + "/configuration.toml"
+
+		err = hooks.CopyFile(srcPath, destPath)
+		if err != nil {
```
same

anonymouse64 commented on hooks/cmd/install/install.go:
```diff
@@ -0,0 +1,360 @@
+
+	// FIXME: chmod 600 pkisetup-<service>.json files; NOT NEEDED???
+
+	err = os.MkdirAll(hooks.SnapData+"/secrets", 0755)
```
should this really be `0755` ? I would imagine for secrets we would want something like `0700`

tonyespy replied:
I'll double check, but again I think this is the same as the existing logic.

anonymouse64 commented on hooks/cmd/install/install.go:
```diff
@@ -0,0 +1,360 @@
+	}
+
+	path = "/security-secret-store/vault-config.hcl"
+	err = hooks.CopyFile(hooks.SnapConf+path+".in", hooks.SnapDataConf+path)
```
nitpick, if this file isn't processed anymore it doesn't need to be named `*.in` I think

anonymouse64 commented on hooks/utils.go:
```diff
@@ -0,0 +1,343 @@
+}
+
+// ModClient returns a testing client
+//func NewTestCtl(g Getter) *Client {
```
commented out code should be removed

anonymouse64 reviewed (approved):
I didn't review the README, nor did I actually test this, but the Go code looks good to me, some minor nitpicks aside (which you can address as a followup or just not at all they don't really matter...)

The one thing I want to call attention to is the order in which we start services, I realize this bug probably existed from my original version of the configure script, but do we need different ordering for stopping / starting services? It seems to me we may have race conditions stopping dependencies before the things that depend on them and introduce undefined behavior if for example postgres is killed right as Kong is performing an ACL lookup, etc.

tonyespy commented on snap/local/runtime-helpers/config/security-secret-store/vault-config.hcl.in:
```diff
@@ -10,11 +10,11 @@
-  tls_client_ca_file ="$SNAP_DATA/secrets/ca/ca.pem"
-  tls_cert_file ="$SNAP_DATA/secrets/edgex-vault/server.crt"
-  tls_key_file = "$SNAP_DATA/secrets/edgex-vault/server.key"
+  cluster_address = ""
```
Because I always go by the rule "change as little as possible". That said, fair point, and maybe I'll add a trello card for exploring further simplification.

tonyespy commented on hooks/cmd/configure/configure.go:
```diff
@@ -0,0 +1,260 @@
+func handleServices() error {
+
+	// grab and log the current service configuration
+	for _, v := range hooks.Services {
```
Probably due to copy and paste from some example that used i (index) and v (value) for the retvals of range.

siggiskulason commented:
@tonyespy - the latest version does not install. The install hook can't run install-setup-kuiper.sh.

```
Apr 25 22:26:54 sidar-XPS-8930 edgexfoundry:configure[195646]: edgexfoundry:install: exec: "install-setup-kuiper.sh": executable file not found in $PATH
Apr 25 22:26:54 sidar-XPS-8930 systemd[1]: snap.edgexfoundry.hook.install.2fb14260-f054-4b10-a450-b8a131d0cf30.scope: Succeeded.
Apr 25 22:26:54 sidar-XPS-8930 snapd[1047]: taskrunner.go:271: [change 27 "Run install hook of \"edgexfoundry\" snap if present" task] failed: run hook "install": exit status 1
```

tonyespy commented:
I just force-pushed and hopefully addressed all of the pending comments. I also updated the related checkbox test based on these changes, and verified that all of the blackbox tests run in both secure and non-secure mode.

Changes made include:
 - modified install-setup-kuiper.sh to executable
 - made more consistent usage of one-line if statements when checking err != nil
 - configure.go:
   - fix iota usage
   - updated on/off constants to ON/OFF
   - re-named handleService(s) functions to better reflect usage
   - use a switch in serviceType function
   - handleAllServices: use s instead of v when iterating services
   - added comments where requested
 - install.go:
   - ensure that $SNAP_DATA/secrets is created with 0700
 - utils.go:
   - improved godoc
   - error out on bad config key (vs. old code which just logged a warning)
 - const.go
   - add constant for clients.command.port

siggiskulason reviewed (approved):
LGTM. Thanks for making those changes and listing them above. I've reviewed them all and they all look fine. I've got not comments. I've also patched my kong user changes on top on your commits and confirmed that it works fine.
